### PR TITLE
Add termination grace period to pod sandbox config

### LINKED ISSUE #77873: Stop timeout isn't respected at shutdown/reboot
**What happened**:
Containers are getting terminated by systemd without respecting the `terminationGracePeriodSeconds` set in the pod yaml on reboot or shutdown of a node.

**What you expected to happen**:
`terminationGracePeriodSeconds` is respected by systemd when using systemd as the cgroup manager. 

**How to reproduce it (as minimally and precisely as possible)**:
1. Use systemd as the cgroup manager in your container runtime.
1. Create a pod yaml with `terminationGracePeriodSeconds` set to `120` seconds.
1. Reboot the node.
1. You will notice that the containers get SIGTERM, followed by systemd default stop timeout (typically 90 seconds) and then they are SIGKILLed.

**Anything else we need to know?**:
This can be fixed by passing the stop timeout to the containers as part of the CreateContainer CRI API. This will allow the container runtimes to set the systemd property for the scope to override the default stop timeout to the value set through `terminationGracePeriodSeconds`.
This needs changes across the stack as runc doesn't currently provide a way to set the TimeoutStopUSec for the systemd scope created for a container.
The behavior with `cgroupfs` cgroup manager will need further investigation.

**Environment**:
- Kubernetes version (use `kubectl version`): All versions.

### PR DESCRIPTION
**Pull request by mrunalp**
This allows us to pass this down to runc and systemd
so systemd can respect this timeout at shutdown.

Signed-off-by: Mrunal Patel <mrunalp@gmail.com>

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
This PR allows us to handle graceful termination of pods at shutdown time. It passes the shutdown
grace period for pods as part of the pod sandbox configuration so it can be passed down to runc and systemd to respect the grace period. 

runc PR for this has been merged  https://github.com/opencontainers/runc/pull/2224

**Which issue(s) this PR fixes**:

Fixes #77873

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
No.

```release-note
Add termination grace period to pod sandox configuration.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

-->
```docs
Add termination grace period to pod sandox configuration.

```

### CODE REVIEW

**Comment by mrunalp** [comment]
@derekwaynecarr ptal

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto [hunk 1]
```diff
@@ -368,6 +368,9 @@ message PodSandboxConfig {
     map<string, string> annotations = 7;
     // Optional configurations specific to Linux hosts.
     LinuxPodSandboxConfig linux = 8;
+    // Grace period in seconds to wait for the containers in a pod to stop before
+    // forcibly terminating them.
+    int64 termination_grace_period = 9;
```
Comment (kolyshkin) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto]: IMHO there's no sense to use negative values here, so maybe `s/int64/uint64/`?

Comment (mrunalp) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto]: Yeah, agree, but my main reason was to match https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L2886.

Comment (kolyshkin) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto]: Got it.

From the top of your head, do you know if there's any special meaning for negative values? If not, I think we need to check and error out either here or in cri-o.

**Comment by mattjmcnaughton** [comment]
cc @Random-Liu who I know has been involved in a lot of CRI stuff in the past.

cc @Joseph-Irving as I know your sidecars implementation interacted w/ termination grace periods a fair amount, so just want to make sure you're aware.

File: pkg/kubelet/kuberuntime/kuberuntime_sandbox.go [hunk 2]
```diff
@@ -22,7 +22,7 @@ import (
 	"net/url"
 	"sort"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (haircommander) [inline pkg/kubelet/kuberuntime/kuberuntime_sandbox.go]: I don't think this change is necessary

Comment (kolyshkin) [reply pkg/kubelet/kuberuntime/kuberuntime_sandbox.go]: This is a side effect of using `gofmt` (or `goexports`) and it's sort of inevitable as golang changes the formatting rules from time to time. Personally, I consider this is a price we have to pay for no coding style wars.

Comment (haircommander) [reply pkg/kubelet/kuberuntime/kuberuntime_sandbox.go]: ah, didn't realize it was linter driven :smile:

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto (hunk 1 above)
Comment (mrunalp) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto]: I have amended the commit to following the minimum grace period that is set for the container stop timeout.

File: pkg/kubelet/kuberuntime/kuberuntime_sandbox.go (hunk 2 above)
Comment (mrunalp) [reply pkg/kubelet/kuberuntime/kuberuntime_sandbox.go]: Yeah, came from there indeed but I have reverted it :)

**Comment by kolyshkin** [comment]
@mrunalp Perhaps we can do without this PR, since `terminationGracePeriod` is passed on to container in annotations, and a CRI can take it from there. This is what I came with for cri-o: https://github.com/cri-o/cri-o/pull/3328

My only concern is label is not standardized and its name can change (it's defined in pkg/kubelet/kuberuntime/labels.go -- maybe we can make it public to avoid copy-paste I do in https://github.com/cri-o/cri-o/pull/3328)

**Comment by mrunalp** [comment]
@kolyshkin Good find :) We can use that to implement this in CRI-O with the updated runc :+1:. However, I still want to keep this open for feedback, because I feel that having this as a first class field in the CRI over an annotation is better.

**Pull request closed without merging**